src/app.tsx

```diff
@@ -0,0 +1,438 @@
+
+interface ProgressBarProps {
+  barLength: number;
+  progressPercent: number;
+}
+
+const ProgressBar: React.FC<ProgressBarProps> = (props) => {
+  const [progressBarLength, setProgresBarLength] = useState(0);
+
+  // runs every time the component rerenders
+  useEffect(() => {
+    setProgresBarLength(props.progressPercent * props.barLength);
+  }, [props.progressPercent]);
+
+  // const progressBarContainerStyle = {
+  //   Position: "absolute",
+  //   left: "100px",
+  //   top: "100px",
+  //   display: "flex",
+  //   alignItems: "center",
+  //   justifyContent: "center",
+  //   ZIndex: "10",
+  // };
+
+  const progressBarStyle = {
+    backgroundColor: "rgb(233, 233, 233)",
+    borderRadius: "0.5rem",
+    width: `${props.barLength}px`,
+  };
+
+  const filledProgressBarStyle = {
+    backgroundColor: "rgb(62, 122, 235)",
+    height: "10px",
+    borderRadius: "1rem",
+    width: `${progressBarLength}px`,
+    // animation
+    transition: "1s ease",
+  };
+
+  return (
+    <div className="absolute top-16 left-16 z-10">
+      <div style={progressBarStyle}>
+        <div style={filledProgressBarStyle} />
+      </div>
+    </div>
+  );
+};
+
+interface GameProps {}
+
+// main parent component
+const Game: React.FC<GameProps> = (props) => {
+  // define states using state hook
+  // typescript implicitly knows the type of each state from initial value
+  const [showModal, setShowModal] = useState(false);
+  const [currentBugType, setCurrentBugType] = useState(""); // so that modal component knows a bug is good or bad
+  const [hasWon, setHasWon] = useState(false);
+  const [correctBugs, setCorrectBugs] = useState(new Set()); // tracking progress
+  const [bugIds, setBugIds] = useState<number[]>([]); // an array of bug ids
+
+  const intervalID = useRef(0);
+
+  // this variable keeps track of total # of spawned bugs
+  // also sets the id for a newest spawned bug
+  const numSpawnedBugs = useRef(0);
+
+  // component did mount
+  useEffect(() => {
+    // create a new bug and give it a unique index
+
+    const spawnNewBug = () => {
+      // if (n > UPPER_COUNT) return
+      if (bugIds.length >= UPPER_BUG_COUNT) {
+        return;
+      } else {
+        numSpawnedBugs.current += 1;
+        setBugIds((prevState) => {
+          return [...prevState, numSpawnedBugs.current];
+        });
+      }
+    };
+
+    intervalID.current = window.setInterval(spawnNewBug, 1000);
+
+    return () => {
+      clearInterval(intervalID.current);
+    };
+  }, [bugIds, numSpawnedBugs]);
+
+  /**
+   * toggles the render state of modal component
+   */
+  const toggleModal = () => {
+    setShowModal((prevState) => !prevState);
+  };
+
+  /**
+   * Delivers info about current clicked bug to Modal and removes the bug from screen
+   *
+   * @param bugId - the id of the bug so that it can be deleted from bugs array
+   */
+  const handleBug = (bugType: string, bugId: number) => {
+    toggleModal();
+
+    setCurrentBugType(bugType);
+
```

Comment:
Catch error early here, so that you can make your long `if-else` a little shorter.

```
if (!BUGS.hasOwnProperty(bugType)) {
   throw new Error("Whoops! The given BugId does not exist in BUGS object.");
}
```

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwnProperty